internal/juju/offers.go

```diff
@@ -390,3 +398,47 @@ func (c offersClient) RemoveRemoteOffer(input *RemoveRemoteOfferInput) []error {
 
 	return nil
 }
+
+// This function adds access to an offer
+func (c offersClient) GrantOffer(input *GrantRevokeOfferInput) error {
+	conn, err := c.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	client := applicationoffers.NewClient(conn)
+	_, err = client.ApplicationOffer(input.OfferURL)
+	if err != nil {
+		return err
+	}
+
+	err = client.GrantOffer(input.User, input.Access, input.OfferURL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// This function revokes access to an offer.
```

Comment:
This function revokes a specific level of access to an offer.

I think (correct me if i'm wrong), but access levels for offers work a bit differently in juju. 
e.g. if a user has `admin` access to an offer and we revoke `admin` access for that user, the user is left with `consume` access to the offer.. if we revoke `consume` access from an `admin` user, that user is left with `read` access. and if we revoke `read` access, user is left with no access to the offer.

Reply:
@alesstimec you are correct in how revoke on offer works.

The provider should decide which level to grant or revoke, not the internal juju code making the API call.

Reply:
@hmlanigan yes, but with the current schema, where juju_access_offer only lists users for a single access level, this is nearly impossible, since you can have a juju_access_offer resource for the `consume` level and another one for `read` level and the two don't "know" about eachother.

Reply:
Also the godoc is not formatted correctly.